util/topgen/c.py

```diff
@@ -93,6 +95,65 @@ def __init__(self, top_info):
         self.top = top_info
         self._top_name = Name(["top", top_info["name"]])
 
+        self._init_plic_mapping()
+
+    def _init_plic_mapping(self):
+        """We eventually want to generate a mapping from interrupt id to the
+        source peripheral.
+
+        In order to do so, we generate two enums, and store the generated names
+        in a dictionary that represents the mapping.
+
+        Plic Interrupt ID 0 corresponds to no interrupt, and so no peripheral,
+        so we encode that in the enum as "unknown".
+
+        The interrupts have to be added in order, with "none" first, to ensure
+        that they get the correct mapping to their PLIC id, which is used for
+        addressing the right registers and bits.
+        """
+        sources = CEnum(self._top_name, Name(["plic", "peripheral"]))
+        interrupts = CEnum(self._top_name, Name(["plic", "irq", "id"]))
+
+        # This will become the final mapping object
+        plic_mapping = {}
+
+        unknown_source = sources.add_constant(Name(["unknown"]), docstring="Unknown Peripheral")
+        none_irq_id = interrupts.add_constant(Name(["none"]), docstring="No Interrupt")
+        plic_mapping[none_irq_id] = unknown_source
+
+        # When we generate the `interrupts` enum, the only info we have about
+        # the source is the module name. We'll use `source_name_map` to map a
+        # short module name to the full name object used for the enum constant.
+        source_name_map = {}
+
+        for name in self.top["interrupt_module"]:
+            source_name = sources.add_constant(Name.from_snake_case(name), docstring=name)
+            source_name_map[name] = source_name
+
+        sources.add_last_constant("Final PLIC peripheral")
+
+        for intr in self.top["interrupt"]:
+            # Some interrupts are multiple bits wide. Here we deal with that by
+            # adding a bit-index suffix
+            if "width" in intr and int(intr["width"]) != 1:
+                for i in range(int(intr["width"])):
+                    name = Name.from_snake_case(intr["name"]).with_more(suffix=[str(i)])
+                    irq_id = interrupts.add_constant(name, docstring="{} {}".format(intr["name"], i))
+                    source_name = source_name_map[intr["module_name"]]
+
+                    plic_mapping[irq_id] = source_name
+            else:
+                name = Name.from_snake_case(intr["name"])
+                irq_id = interrupts.add_constant(name, docstring=intr["name"])
+                source_name = source_name_map[intr["module_name"]]
+                plic_mapping[irq_id] = source_name
```

Comment:
To avoid the duplicated code, you could do something like this:
```python
        for intr in self.top["interrupt"]:
            # Some interrupts are multiple bits wide. Here we deal with that by
            # adding a bit-index suffix
            base_name = Name.from_snake_case(intr["name"])
            width = int(intr.get("width", 1))
            if width != 1:
                pairs = [(base_name.with_more(suffix=[str(i)]),
                          "{} {}".format(intr["name"], i))
                         for i in range(width)]
            else:
                pairs = [(base_name, intr["name"])]

            for name, docstring in pairs:
                irq_id = interrupts.add_constant(name, docstring=docstring)
                source_name = source_name_map[intr["module_name"]]
                plic_mapping[irq_id] = source_name
```

If you rejig the `CEnum` constructor like I suggested above, this mostly Just Works. The only problem is constructing `plic_mapping`. You could do something like this:

```python
        intr_pairs = []
        source_names = []

        for intr in self.top["interrupt"]:
            # Some interrupts are multiple bits wide. Here we deal with that by
            # adding a bit-index suffix
            base_name = Name.from_snake_case(intr["name"])
            width = int(intr.get("width", 1))
            if width != 1:
                intr_pairs += [(base_name.with_more(suffix=[str(i)]),
                                "{} {}".format(intr["name"], i))
                               for i in range(width)]
            else:
                intr_pairs.append((base_name, intr["name"]))

            source_names += [source_name_map[intr['module_name']]] * width

        ...

        interrupts = CEnum(self._top_name.with_more(suffix=['plic', 'irq', 'id']),
                           intr_pairs,
                           'The Last Valid Interrupt ID.')

        ...

        assert len(interrupts.constants) == len(source_names)
        for (irq_id, _, _), source_name in zip(interrupts.constants, source_names):
            plic_mapping[irq_id] = source_name
```

Reply:
I really think doing a post-hoc `zip` is playing with fire here, when there is a point in the code when we do have the two mapping entries together.

Reply:
I guess it's a style preference. No strong feelings here.